DOC: remove references to RandomState

tupui opened the pull request:
## Status
Since NumPy 1.17, the [canonical way](https://numpy.org/doc/stable/reference/random/index.html) to create random numbers is:

```python
import numpy as np
from numpy.random import default_rng
rng = default_rng()
```
And If a seed is to be used. User should not use a common value such as 0 or 123456. This is to prevent bias (everyone is using the same values) and ensure sufficient entropy for the generator. More [details](https://numpy.org/doc/stable/reference/random/bit_generators/index.html#seeding-and-entropy). The good practice (from NumPy's doc) is:

```python
from numpy.random import SeedSequence
seed = SeedSequence().entropy
```

But in our documentation, we present the old way of dealing with RNG and show bad things like global states of RNG.
## Proposal
I propose here to remove from the documentation any reference to `RandomState` and wrong (not best practice) usage of random number generators. It implies changing docstrings `random_state,seed` and using correct RNG in examples and tutorial. The main user is stats and the tutorial has been updated with a dedicated section for random number generation.

For reference, a discussion we recently had about RNG and seeding here: https://github.com/scipy/scipy/pull/13631#discussion_r598937552

ilayn commented:
Well it's not wrong but just the old and not optimal way of doing it. Including the majority of our test suite, there is a wild amount of `seed` usage so I wouldn't jump to that conclusion.

tupui commented:
> Well it's not wrong but just the old and not optimal way of doing it. Including the majority of our test suite, there is a wild amount of `seed` usage so I wouldn't jump to that conclusion.

Using `RandomState` is not "wrong" but should not be used by final users. It is explicitly written in NumPy's doc that it stays for compatibility reasons. But the generator is not optimal (only one option left in new versions). 

On the test side, yes we have to keep it for now as this is the only way now which is safely providing a deterministic stream of bits accros platforms and versions. Also noted in NumPy's doc.

IMO, we should only show good practices. From my experience, practitioners take our doc for a gold standard and in that regards, we have to be careful about what we show in examples. A lot of people just copy paste code and don't touch anything if it works. So if in examples we showcase something using a bad, or even just old, way it is not good.

As for taking a common seed and using a global seed. It is definitely an issue.

ilayn commented:
I agree with all but we have to be careful by swinging from one side to the other. A gentler approach will go a long way for convincing users. Not changing every version to another conclusion. 1.17 was wrong 1.20 was correct now also wrong 1.34 is definitely correct now is not emitting the right message here. Showing the shortcomings and letting them convince themselves is a more effective strategy.

rkern commented on doc/source/tutorial/stats.rst:
```diff
@@ -178,6 +162,39 @@ keyword argument, ``loc``, which is the first of a pair of keyword arguments
+And fixing the seed can be done like this:
+
+    >>> from numpy.random import SeedSequence
+    >>> seed = SeedSequence().entropy
```
This example doesn't fix the seed. Rerunning the whole example gets fresh entropy every time. Instead, I would put in an explicit number literal, like the old example. However, to improve on the old example, use a quality 128-bit number and in the `.. warning::` below show how you can use `SeedSequence().entropy` to get such a number.

tupui replied:
Ok thanks. I was really how to deal with this one. I should have said here that you could get a number by running the code and then use it as you pleased.

Ideally I did not want to put any number so that users would not copy past the same number all the time. I also thought about generating a fix number which would change every time you reload the page.

rkern commented on scipy/linalg/_sketches.py:
```diff
@@ -26,12 +26,12 @@ def cwt_matrix(n_rows, n_columns, seed=None):
-        If integer, it is used to seed the local ``RandomState`` instance.
-        Default is None.
+    seed : {None, int, `numpy.random.Generator`}, optional
+        If `seed` is None the `numpy.random.Generator` singleton is used.
```
There is no `Generator` singleton. The default behavior of the `check_random_state()` function that this method uses will still use the `RandomState` singleton when `seed is None`. Similarly, when `seed` is an integer, you will get a `RandomState` instance, not a `Generator`.

The old version in the `rvs()` method down below is the most accurate description. Basically, old versions that don't mention `Generator` should be updated to include that option, but these new versions that omit references to `RandomState` entirely are currently incorrect.

rkern commented on scipy/sparse/construct.py:
```diff
@@ -756,12 +762,12 @@ def random(m, n, density=0.01, format='coo', dtype=None,
     ...     def _rvs(self,  size=None, random_state=None):
-    ...         return random_state.randn(*size)
-    >>> X = CustomDistribution(seed=2906)
+    ...         return random_state.random(*size)
```
`standard_normal()` would be the replacement for `randn()`.

tupui replied:
oh yes, thanks!

rkern commented on scipy/sparse/construct.py:
```diff
@@ -738,12 +742,14 @@ def random(m, n, density=0.01, format='coo', dtype=None,
+    >>> rng = default_rng()
 
-    >>> class CustomRandomState(np.random.RandomState):
+    >>> class CustomRandomState(np.random.Generator):
```
How necessary is this subclassing to the example? It's not commented on, or described as an intended API elsewhere in the docstring. Changing the semantics of a method like this is not a practice that I would recommend. If one needs to be able to specify to this method to sample even integers, as in this example, that needs to be specified by some other means.

Also, just swapping `Generator` for `RandomState` won't work in this example. This function uses the `rng_integers()` compatibility shim to draw integers. If `random_state` is a `Generator`, it will call the `integers()` method directly instead of `randint()`, so this overridden method gets ignored.

tupui replied:
Indeed, I don't think this class is needed in the example. I will remove it.

rkern commented on scipy/spatial/transform/rotation.pyi:
```diff
@@ -39,7 +42,7 @@ class Rotation:
     @classmethod
     def random(cls, num: Optional[int] = ...,
-               random_state: Union[int, np.random.RandomState, None] = ...) -> Rotation: ...
+               random_state: Optional[Union[_IntegerType, np.random.Generator]] = ...) -> Rotation: ...
```
`RandomState` is still an option, as is `None`.

rkern replied:
Ah, `Optional` accounts for the `None`. This old-timer is still getting used to the type annotations.

rkern commented on scipy/spatial/transform/rotation.pyx:
```diff
@@ -2169,12 +2171,12 @@ cdef class Rotation(object):
         Sample a stack of rotations:
 
-        >>> R.random(5, random_state=1234).as_euler('zxy', degrees=True)
+        >>> R.random(5).as_euler('zxy', degrees=True)  # doctest: +SKIP
```
Why not annotate with `# random`?

tupui replied:
I was not sure it worked, thanks I will use this then.

rkern commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -1036,13 +1035,12 @@ def rvs(self, *args, **kwds):
+            If `seed` is an int, a new ``Generator`` instance is used,
+            seeded with `seed`.
+            If `seed` is already a ``Generator`` instance then that instance is
+            used.
```
The old version is the most correct.

rkern commented on scipy/stats/_multivariate.py:
```diff
@@ -2063,7 +2056,12 @@ def _standard_rvs(self, n, shape, dim, df, random_state):
+            If `seed` is an int, a new ``Generator`` instance is used,
+            seeded with `seed`.
+            If `seed` is already a ``Generator`` instance then that instance is
+            used.
```
The old version is more correct. By the time this method gets called, the value has already been coerced into either a `RandomState` or a `Generator`.

rkern reviewed (commented):
Changing the examples to use `default_rng()` instead of `np.random` or `RandomState` are very welcome. Some of the docstring changes for `random_state` arguments need to be reverted to accurately describe the behavior of `check_random_state()`.

tupui commented:
Thanks @rkern for the review.

As you see I wanted to get rid of all `RandomState` even if under the hood we are calling `check_random_state`. We had a discussion when working on QMC and the result was that I coded a variant of this (`scipy.stats._qmc:check_random_state`). In the docstring of the module, we do not talk at all about `RandomState` even if yes, the functions would accept it. This is to push the user to use the Generator interface.

The idea here was to do the same everywhere. Keep this hidden behaviour while encouraging user to use RandomState. If we keep RandomState as options in the arguments, the linters will be fine. Which is not what I wanted.

andyfaff commented:
I'm in agreement with @rkern here. I'd prefer the docstring for the parameters to reflect what's actually accepted by the function. For the examples, changing them all to use `np.random.Generator` is fine.

Half of me does wonder if a phased deprecation away from RandomState would be worthwhile, but that would need to be part of a wider conversation. It would make internal code easier.

rkern commented:
Omitting the `RandomState` option when describing the `seed` arguments in the QMC code makes sense; it's an incidental legacy option, and repeating the full list of options each time is overly verbose. I do advocate for not having to describe every possible valid value when some of them are not recommended. You are describing the current semantics of the code, just being judicious with the limited amount of text we want to spend describing it.

But we're not using that version of `check_random_state()` in all of these places. `RandomState` is not incidental to the processing of these arguments. Rather, it's fundamental to the processing of every valid input _except_ a full-fledged `Generator` instance. Omitting mention of `RandomState` for those valid inputs means telling the user something incorrect about the current version of the code, not just omitting a valid but disrecommended option. It would be describing a possible future state of the code, but we haven't decided when (or whether, in all particulars) we'll commit to those semantics.

mdhaber commented:
I see now that `Generator` does not have `randint`. What are we _supposed_ to do if we want to write a function that works with both `RandomState` and `Generator`? (Do we really need to try/catch or detect what type of object it is and call the appropriate method?)

andyfaff commented:
> I see now that Generator does not have randint, and RandomState does not have integers. What are we supposed to do if we want to write a function that works with both RandomState and Generator? (Do we really need to try/catch or detect what type of object it is and call the appropriate method?)

Well, funny you should ask this question, there is functionality for this [here](https://github.com/scipy/scipy/blob/master/scipy/_lib/_util.py#L449). This was introduced in #11680.

It might be better to write a wrapper:

```
def rng_integers(gen, low, high=None, size=None, dtype='int64',
                 endpoint=False):
    rng = check_random_state(gen)
    
    if isinstance(rng, np.random.Generator):
        def gener(low, high=None, size=None, dtype='int64', endpoint=False)
            return rng.integers(low, high=high, size=size, dtype=dtype,
                                endpoint=endpoint)
    else:
        # RandomState
        def gener(low, high=None, size=None, dtype='int64', endpoint=False)
            if endpoint:
                # inclusive of endpoint
                # remember that low and high can be arrays, so don't modify in
                # place
                if high is None:
                    return rng.randint(low + 1, size=size, dtype=dtype)
                if high is not None:
                    return rng.randint(low, high=high + 1, size=size, dtype=dtype)

            # exclusive
            return rng.randint(low, high=high, size=size, dtype=dtype)
    
    return gener
```

mdhaber commented:
Thanks @andyfaff! I think I'll be ok with the function as-is.

tupui commented:
OK then I will see to add back `RandomState` where we are using `check_random_state`.

@andyfaff I am really in favour of planning a deprecation cycle for `RandomState`. We are almost in a PY2-PY3 situation here.
There is a huge legacy here and we must be careful, I get this. On the other hand, some will never change because it "still" works (cf. the recent small discussion on the mail list). But for the sake of users not willing to update some values in their tests, we are paying a high dev price. (Yes, I saw production code with seeds too... But IMO this is just another wrong practice and hence should not be of any concern on our side). Even for the user it's confusing to see that many options for a given parameter. Sometimes you have to break things for people to move on and adopt best practices.

Also, feel free to commit directly on the branch if you see something.

mdhaber commented:
@tupui Can you point us to the TLDR of why it matters? I wonder if I'm the only maintainer who doesn't know. I'm not arguing; I believe that it's the "right" thing to do now, but I haven't come across any answers that resonate with me as to why it matters in everyday code. The documentation says:
```
The main difference between the two is that Generator relies on an additional BitGenerator to manage state and generate the random bits, which are then transformed into random values from useful distributions.
```
but this doesn't state what the _difference_ is or why it's good; it only states what `Generator` does.

tupui commented:
> @tupui Can you point us to the TLDR of why it matters? I wonder if I'm the only maintainer who doesn't know. I'm not arguing; I believe that it's the "right" thing to do now, but I haven't come across any answers that resonate with me as to why it matters in everyday code. 

TBH, in terms of statistical properties of the RNG, I don't know as I did not dive into it. But on the practical side, there are new things like support of `axis` for `choice, permutation, shuffle`. Or just a clearer interface as there are not anymore several functions doing the same (or almost) things like `random_sample` and `rand`. Also if my reading is correct, the new generators are faster.

But @rkern would answer better here as he wrote NEP19 https://numpy.org/neps/nep-0019-rng-policy.html

mdhaber commented:
Thanks, I'll read that.

andyfaff commented:
The NEP is a nice read. That definitely suggests to me the worth the deprecation cycle to remove the technical debt.

tupui commented:
> The NEP is a nice read. That definitely suggests to me the worth the deprecation cycle to remove the technical debt.

If we would go with this, I could also change the PR toward deprecating it. Let me all know 😄

rgommers commented:
> If we would go with this, I could also change the PR toward deprecating it.

Please don't include it here. I don't think we want to do that deprecation - that's a large amount of churn for users, for limited benefits for SciPy itself. But if we want to do it, it's best to discuss on a separate issue/PR.

rkern commented:
There are a couple of things that might people think might be meant by "deprecating it". The one I think some are arguing against is to noisily warn about and eventually reject `RandomState` instances that are passed in.

However, I do think at some point we will want to change the behavior of the general `check_random_state()` to be like the version in the QMC code, to be `Generator`-first, processing integer and `None` seeds with `default_rng()` instead of `RandomState` (and probably also extracting the `BitGenerator` from a passed-in `RandomState` to wrap with `Generator` instead of passing the `RandomState` through). We don't fix bugs in the distributions on `RandomState` anymore, and we will often want to work with the newer APIs that are only available on `Generator`. Those changes probably do warrant some kind of managed change, though I'm not convinced it has to be noisy deprecation warnings.

rkern commented:
But yeah, not in this branch. The example cleanups (and mentioning `Generator` in the docstrings that hadn't gotten that update) are a great contribution alone.

tupui commented:
@rkern can you just confirm that I can use this docstring for seed using the current `check_random_state`:

```python
    seed : {None, int, `numpy.random.Generator`,
            `numpy.random.RandomState`}, optional
        If `seed` is None (or `np.random`), the `numpy.random.RandomState`
        singleton is used.
        If `seed` is an int, a new ``RandomState`` instance is used,
        seeded with `seed`.
        If `seed` is already a ``Generator`` or ``RandomState`` instance then
        that instance is used.
```

This is just a slightly modified version of the one from QMC.

ev-br commented:
One possible enhancement (in a separate PR!) could be to make `check_random_state` recognize a `np.random.SeedSequence` instance and seed a `Generator` with it.

tupui commented:
> One possible enhancement (in a separate PR!) could be to make `check_random_state` recognize a `np.random.SeedSequence` instance and seed a `Generator` with it.

I can do this afterwards sure 😉

tupui commented:
I updated the PR with our latest discussions. Let me know what you think.

Side note: I added a commit fixing some small pep8. Only pep8 things are in this commit. I can revert this in case you don't like it. But as we don't do PR just for this, I though it could be nice while I was going through these files.

rkern commented on scipy/optimize/_dual_annealing.py:
```diff
@@ -38,9 +38,16 @@ class VisitingDistribution:
-        for using the current state of the created random generator container.
+    rand_gen : {None, int, `numpy.random.Generator`,
+                `numpy.random.RandomState`}, optional
+
```
This argument is not coerced via `check_random_state()`. This class and and `StrategyChain` are used internally after the `seed=` argument has been coerced already.

tupui replied:
OK, I though of doing it this way because `check_random_state` would still feed a `RandomState` to this class. I will update it then.

rkern replied:
Sorry, let me clarify: the old version of this docstring was correct and should just be reverted. The value is either a `RandomState` or a `Generator`, but it cannot be `None` or `int`. `VisitingDistribution` and `StrategyChain` are internal classes that will only be called with an actual `RandomState` or `Generator` instance. The user will call `dual_annealing()` with a `seed` argument; `dual_annealing()` is responsible for calling `rand_gen = check_random_state(seed)` and will then pass `rand_gen` (which is then only either a `RandomState` or `Generator` instance) to `VisitingDistribution` and `StrategyChain`.

tupui replied:
Oh right, I fixed it.

rkern commented on scipy/optimize/_lsq/least_squares.py:
```diff
@@ -667,12 +667,14 @@ def least_squares(
+    >>> import numpy as np
+    >>> from numpy.random import default_rng
+    >>> rng = default_rng()
+    >>> def gen_data(t, a, b, c, noise=0., n_outliers=0, random_state=None):
```
Put the `default_rng()` call inside the function. Rename the `random_state` argument to `seed` (for readability's sake), and do `rng = default_rng(seed)` inside of the function.

rkern commented on scipy/stats/stats.py:
```diff
@@ -5193,10 +5173,12 @@ def multiscale_graphcorr(x, y, compute_distance=_euclidean_dist, reps=1000,
-        If seed is an int, return a new RandomState instance seeded with seed.
-        If None, use np.random.RandomState. Default is None.
+    random_state : {None, int, `numpy.random.Generator`}, optional
+        If `seed` is None the `numpy.random.Generator` singleton is used.
```
This and the other arguments below in this file still need to be fixed to the latest version.

tupui replied:
Indeed, sorry I missed that. Fixing, this file is so long...

rkern reviewed (commented):
A few more fixes.

I'm neutral on the PEP8ifying.

ilayn commented on scipy/linalg/decomp_svd.py:
```diff
@@ -254,6 +257,7 @@ def diagsvd(s, M, N):
 
     Examples
     --------
+    >>> import numpy as np
```
This line is implicit to all examples in SciPy hence we omit it.

tupui replied:
Ok thanks I will fix all of these.

ilayn commented on scipy/linalg/decomp_svd.py:
```diff
@@ -306,6 +310,7 @@ def orth(A, rcond=None):
 
     Examples
     --------
+    >>> import numpy as np
```
Same here

ilayn commented on scipy/linalg/decomp_svd.py:
```diff
@@ -355,6 +360,7 @@ def null_space(A, rcond=None):
     --------
     1-D null space:
 
+    >>> import numpy as np
```
Same here

ilayn commented on scipy/linalg/decomp_svd.py:
```diff
@@ -429,7 +437,10 @@ def subspace_angles(A, B):
     An Hadamard matrix, which has orthogonal columns, so we expect that
     the suspace angle to be :math:`\frac{\pi}{2}`:
 
+    >>> import numpy as np
```
Same here

ilayn commented on scipy/optimize/_lsq/least_squares.py:
```diff
@@ -667,12 +667,16 @@ def least_squares(
     outliers, define the model parameters, and generate data:
 
-    >>> def gen_data(t, a, b, c, noise=0, n_outliers=0, random_state=0):
+    >>> import numpy as np
```
Same here

ilayn commented on scipy/optimize/_qap.py:
```diff
@@ -348,13 +347,15 @@ def _quadratic_assignment_faq(A, B,
     solution than a single random initialization.
 
-    >>> np.random.seed(0)
+    >>> import numpy as np
```
same here

ilayn commented on scipy/signal/signaltools.py:
```diff
@@ -3428,9 +3429,11 @@ def detrend(data, axis=-1, type='linear', bp=0, overwrite_data=False):
     --------
     >>> from scipy import signal
-    >>> randgen = np.random.RandomState(9)
+    >>> import numpy as np
```
same here

ilayn commented on scipy/sparse/construct.py:
```diff
@@ -738,12 +745,14 @@ def random(m, n, density=0.01, format='coo', dtype=None,
     --------
     >>> from scipy.sparse import random
     >>> from scipy import stats
+    >>> import numpy as np
```
same here

ilayn commented on scipy/stats/stats.py:
```diff
@@ -2251,7 +2229,10 @@ def cumfreq(a, numbins=10, defaultreallimits=None, weights=None):
     Examples
     --------
     >>> import matplotlib.pyplot as plt
+    >>> import numpy as np
```
same here

ilayn commented on scipy/stats/stats.py:
```diff
@@ -2333,7 +2312,10 @@ def relfreq(a, numbins=10, defaultreallimits=None, weights=None):
     Examples
     --------
     >>> import matplotlib.pyplot as plt
+    >>> import numpy as np
```
same here

tupui commented:
@rkern @ilayn I integrated your suggestions and fixed the CI. Could you please have another look?

rkern commented on doc/source/tutorial/stats.rst:
```diff
@@ -152,32 +152,53 @@ argument:
+Drawing random numbers relies on generators from `numpy.random` package.
+In the examples above, the specific stream of
+random numbers is not reproducible across runs. To achieve reproducibility,
+you can explicitly *seed* the random number generator. In NumPy, a generator
```
```suggestion
you can explicitly *seed* a random number generator. In NumPy, a generator
```

rkern commented on doc/source/tutorial/stats.rst:
```diff
@@ -152,32 +152,53 @@ argument:
+.. warning:: Do not use this number or common values such as 0. Using just a
+             small set of seeds to instantiate larger state spaces means that
+             there are some initial states that are impossible to reach. This
+             creates some biases if everyone uses such values. A proper way to
```
```suggestion
             creates some biases if everyone uses such values. A good way to
```

I prefer this softening.

rkern commented on scipy/linalg/decomp_svd.py:
```diff
@@ -73,8 +73,10 @@ def svd(a, full_matrices=True, compute_uv=True, overwrite_a=False,
     --------
     >>> from scipy import linalg
+    >>> from numpy.random import default_rng
+    >>> rng = default_rng()
```
One more possible tweak is to omit the import and use `np.random.default_rng()` in the examples. Up to you.

tupui replied:
I think it would make the line a bit long then. If you don't mind I will let this like that.

rkern reviewed (approved):
Just a few more very minor suggestions to consider.

tupui commented:
Thank you for the review @rkern! Shall we merge then?

rkern commented:
@mdhaber
> Do you happen to know if this was an oversight (i.e. that people might want to use `RandomState` and `Generator` the same way) or was the name `randint` deemed too egregious to keep around?

The `randint()` API had accumulated a lot of cruft as we tacked on functionality and fixed deficiencies of the original implementation (e.g. to correctly handle the semantics of the non-inclusive upper bound for each dtype even though that bound cannot be expressed in that dtype) while maintaining backwards compatibility in `RandomState`. `integers()` is the cleaned-up version, how we would design the API given the current set of functionality if we had a fresh start. Because the API is different, it needed a different name. If it were solely up to me, I might still have kept the `randint()` method there (translating its calls to `integers()` internally) so that `Generator` would be more of a drop-in replacement, but using `rng_integers()` (which mimics the `integers()` interface and translates it to `randint()` when `RandomState` is used) is a better in the long-term as `RandomState` disappears. We won't keep the legacy of the crufty `randint()` interface forever.

rkern commented:
@tupui This final form LGTM.

tupui commented:
> @tupui This final form LGTM.

Thanks. I propose to merge this tomorrow then.

mdhaber commented:
@rkern thanks. I hadn't noticed the API difference yet. `rng_integers` is a good bridge during the transition.

sturlamolden commented:
> ## Status
> Since NumPy 1.17, the [canonical way](https://numpy.org/doc/stable/reference/random/index.html) to create random numbers is:
> 
> ```python
> import numpy as np
> from numpy.random import default_rng
> rng = default_rng()
> ```

NumPy documentation says that the default generator can change in future version.

“The Generator is the user-facing object that is nearly identical to the legacy RandomState. It accepts a bit generator instance as an argument. The default is currently PCG64 but this may change in future versions.”

https://numpy.org/doc/stable/reference/random/index.html

This “canonical” way of producing random variates can therefore break the test suite and the doctests. Generator and seed should be explicitly specified.

tupui commented:
> This “canonical” way of producing random variates can therefore break the test suite and the doctests. Generator and seed should be explicitly specified.

Robert already replied in the PR.

On my side, I stand by this change. It would be totally up to me that there would be absolutely no seeding. It can give you a fake sense of things working although they are not. 

Anyway, here the examples are user facing and yes NEP19 is explicitly advocating for using what we proposed here. It's a very little prize to pay to fix a few numbers in the unlikely even that the generator changes. FYI, it took me at most 1h to replace them all once I had the logic in place. So not really any discussions to have here. Furthermore, the tests are still using `RandomState` as NEP19 promotes.

Good hygiene around RNG is paramount and as a pilar of the statistical environment in Python, SciPy must be exemplary.
